src/lib.rs

```diff
@@ -859,6 +852,83 @@ pub fn wasm2wat<S: AsRef<[u8]>>(wasm: S) -> Result<String, Error> {
     Ok(text)
 }
 
+struct WabtWriteScriptResult {
+    raw_script_result: *mut ffi::WabtWriteScriptResult,
+}
+
+struct WabtWriteScriptResultRelease {
+    json_output_buffer: WabtBuf,
+    _log_output_buffer: WabtBuf,
+    module_output_buffers: Vec<(String, WabtBuf)>,
+}
+
+impl WabtWriteScriptResult {
+    fn is_ok(&self) -> bool {
+        unsafe {
+            ffi::wabt_write_script_result_get_result(self.raw_script_result) == ffi::Result::Ok
+        }
+    }
+
+    fn module_count(&self) -> usize {
+        unsafe {
+            ffi::wabt_write_script_result_get_module_count(self.raw_script_result)
+        }
+    }
+
+    fn module_filename(&self, index: usize) -> &str {
+        assert!(index < self.module_count());
+        unsafe {
+            use std::ffi::CStr;
+
+            let s = ffi::wabt_write_script_result_get_module_filename(self.raw_script_result, index);
+            let s = CStr::from_ptr(s);
+            s.to_str().unwrap()
+        }
+    }
+
+    fn take_all(self) -> Result<WabtWriteScriptResultRelease, ()> {
+        if self.is_ok() {
+            let json_output_buffer;
+            let log_output_buffer;
+            let mut module_output_buffers = Vec::new();
+            unsafe {
+                json_output_buffer =
+                    ffi::wabt_write_script_result_release_json_output_buffer(
+                        self.raw_script_result);
+                log_output_buffer =
+                    ffi::wabt_write_script_result_release_log_output_buffer(
+                        self.raw_script_result);
+            }
+            for i in 0..self.module_count() {
+                let module_output_buffer = unsafe {
+                    ffi::wabt_write_script_result_release_module_output_buffer(
+                        self.raw_script_result, i)
+                };
+                let name = self.module_filename(i);
```

Comment:
Can we instead of `CStr` use `String::from_utf8_lossy`? At least it doesn't panic upon issues with utf8

Reply:
Hm, I found using the `lossy`  variants dangerous for this, as you might generate name collisions that way, and I was under the assumption that WebAssembly embraced utf8 anyway. 

I could also just store the names as `CString` or `Vec<u8>` there?

Reply:
That's actually a good idea! We are only using these filenames for matching them with JSON commands, right? Even if we also need these filenames in order to display error messages, I think we can use `from_utf8_lossy` here (since error reporting isn't that critcial and it doesn't panic).